packages/liquidator/mocha-test/liquidationStrategy.js

```diff
@@ -50,6 +51,62 @@ describe("LiquidatorStrategy", () => {
     result = strat.utils.withdrawProgressPercent(1000, 2000, 1500);
     assert.equal(result.toFixed(2), "50.00");
   });
+  it("hasWithdrawRequestPending", () => {
+    const config = {
+      minSponsorSize: 10
+    };
+    let strat = Strategy(config, { toBN, BN });
+
+    // Withdraw pending
+    let result = strat.utils.hasWithdrawRequestPending({
+      position: { withdrawalRequestPassTimestamp: 11 },
+      currentBlockTime: 10
+    });
+    assert(result);
+    // No withdraw pending
+    result = strat.utils.hasWithdrawRequestPending({
+      position: { withdrawalRequestPassTimestamp: 0 },
+      currentBlockTime: 10
+    });
+    assert(!result);
+    // Withdraw expired
+    result = strat.utils.hasWithdrawRequestPending({
+      position: { withdrawalRequestPassTimestamp: 9 },
+      currentBlockTime: 10
+    });
+    assert(!result);
+  });
+  it("passedDefenseActivationPercent", () => {
+    let config = {
+      minSponsorSize: 10,
+      defenseActivationPercent: 80,
+      withdrawLiveness: 100
+    };
+    let strat = Strategy(config, { toBN, BN });
+
+    // Withdrawal request expires at 1000, liveness = 100,
+    // current block = 900, defense activation = 0%
+    // ==> FALSE
+    let result = strat.utils.passedDefenseActivationPercent({
+      position: { withdrawalRequestPassTimestamp: 1000 },
+      currentBlockTime: 900
+    });
+    assert(!result);
+    // current block = 950, defense activation = 50%
+    // ==> FALSE
+    result = strat.utils.passedDefenseActivationPercent({
```

Comment:
I think I have a handle on all of this except for the defense activation percent. Could you explain what this is and how it's used in the context of the whale defense fund strategy?

Reply:
When a withdrawal is requested it has a `liveness` period during which it cannot be executed. If a position with a pending withdrawal is liquidated, then its liveness gets extended. For example, if a liveness period is 100 seconds long, and 20 seconds have elapsed when a liquidation occurs, then another 100 seconds gets added to the withdrawal's liveness. In this example, the 80 seconds remaining get another 100 seconds for a total of 180 seconds of liveness.

The `defenseActivatioPercent` controls when liquidations are sent, so that bots can efficiently extend withdrawal periods. If the `defenseActivationPercent` is 50%, then a liquidation is only sent when at least 50% of the withdrawal's liveness has expired. Using the previous example of 100 second withdrawal liveness, the first liquidation can be submitted after 50 seconds have elapsed. The new withdrawal liveness will be 50 + 100 = 150 seconds. The next liquidation will occur after 75 seconds. And so on.